New slinky integration

oyvindberg opened the pull request.

oyvindberg committed:
Downgrade sbt so intellij can resolve sources again

oyvindberg committed:
Import code for antd-slinky demo, start sketching new slinky integration

mcallisto commented:
Is there a reason why App is not as Slinky-fied as it could be?
* @react annotation
* Hooks
* web.html components

oyvindberg commented:
No, not much thought have gone into it. I thought it was pretty close to yours, but if anything is missing/different feel free to add it - you should have write access here now.
The only thing i knowingly left out was your css, and that was just laziness. feel free to re-add that as well

mcallisto commented:
A more precise question. Since you said:

> i think it would be nice to have a few react demos done in a duo/trio with only scalablytyped, with slinky, and possible with scalajs-react

This is the _only scalablytyped_.
If I Slinky-fy it, it becomes the one _with slinky_.

So how you have imagined the duo?
In the meanwhile I will take care of the css.

At the same time, thank you for all the lessons. I start to see that doing it with or without Slinky is more or less the same with no so great differences.

mcallisto commented:
Ok forget, this last comment, I now see that there is an **antd-slinky** demo!

mcallisto committed:
Added CSS for proper rendering

mcallisto committed:
Added CSS in antd-slinky as well

mcallisto committed:
Minor Slinky additions

mcallisto commented:
It seems good to go.

Maybe you want first to publish the improved `react-slinky-facade` so that I can be added as an external dependency.

Just a minor point. In the slinky version, I moved from
`val noteTitle: js.UndefOr[String]`
to
`val noteTitle: String`
My idea is that you want less `js` around as possible and also you should know the props needed by the component you intend to use.
Or maybe we could have `val noteTitle: String = "Note"` with a given default.

mcallisto committed:
react-facade not needed in antd-slinky

mcallisto committed:
Added react-router-dom-slinky demo

oyvindberg commented:
Awesome stuff @mcallisto. Im afk until tomorrow late, so will have a better look then. 
Will definitely publish the new facade before we merge this, though I've hoped to experiment a bit and see what other improvements can be found :)

oyvindberg committed:
Proposal for what Slinky facades can look like.

This involves a manual layer on top of ScalablyTyped which exports pretty much everything a user might need:
- prop types/creators unchanged
- react components as slinky `ExternalComponent` types
- custom code to make complicated things easier

oyvindberg commented:
Everything you pushed looked good @mcallisto . I gave it a go to make a facade for `Route`, WDYT?

oyvindberg commented:
Ok, I had a bit more time to flesh out how I imagine a slinky facade layer can look like. Take a look at `AntdFacade` and `ReactRouterFacade` in the newest commit, and see how they are used. I can describe these ideas in more text if it's not clear, but some input on this would be very very useful now :)

mcallisto committed:
Better title

mcallisto commented:
Really positively impressed.

I think the two layers solution (`ScalableSlinky` + individual library facade) is great.
* encourages others to try writing facades, they have a blueprint
* makes the final code clearer and cleaner, short and in accordance with Slinky "dialect"

mcallisto commented on antd-slinky/src/main/scala/demo/AntdFacade.scala:
```diff
@@ -0,0 +1,88 @@
+  def formComponent[Props <: js.Object](
+      options: FormCreateOption[Props]
+  )(f:         js.Function1[Props with WithForm, ReactElement]): ExternalComponentP[Props] =
+    importSTComponent(createForm(FormCreateOption[Props](name = "coordinated"))(f).asInstanceOf[ComponentType[Props]])
```
I wonder if `name` shouldn't be a param of  `def formComponent`, to avoid conflicts between multiple forms (in the same page?)

oyvindberg replied:
Not sure. `FormCreateOption` has a few other things, i have no idea if they are useful

mcallisto replied:
Ok, you are right, but then in this case in the function the `options` param must be utilized, I changed this in https://github.com/oyvindberg/ScalablyTypedDemos/pull/10/commits/8a116ed8bb3b1e0b144989524a4c0e998d980fa9

oyvindberg commented:
Happy you like it so far @mcallisto . I'll let this stay open for some more days and think about what else we can do to make the situation better :)

mcallisto committed:
Use options param

mcallisto committed:
Added Notification example

oyvindberg committed:
- Use props traits for forwarding props creators in slinky facades
- Use  `toFuture` on `js.Promise`

oyvindberg commented:
Just a headsup: I just pushed a commit which wont build until we update it with the new ScalablyTyped build in the morning

We no longer need to write all the props forwarders, since i put them in a trait :)

oyvindberg committed:
Handle failed requests with both axios and fetch

oyvindberg commented:
I [merged](https://github.com/oyvindberg/ScalablyTyped/commit/c327a81397c01fd8e640a5947561d4849768fe13) the slinky integration now, so it'll be released tomorrow. merging this after that.

oyvindberg committed:
Lift antd and react-router-dom slinky facades up to ScalablyTyped
